hal_st/stm32fxxx/DmaStm.hpp

```diff
@@ -114,18 +115,23 @@ namespace hal
             uint8_t dmaMux;
 #endif
             uint8_t streamIndex = 0xff;
-   
+
 #ifdef GPDMA1
             struct LinkedList
             {
-                uint32_t CTR1;
-                uint32_t CTR2;
                 uint32_t CBR1;
                 uint32_t CSAR;
                 uint32_t CDAR;
                 uint32_t CLLR;
             };
-            LinkedList linkRegisters;
+
+            std::array<LinkedList, 8> linkMemoryArrayDma1{};
+#if defined(DMA2_Channel1) || defined(GPDMA2)
+            std::array<LinkedList, 8> linkMemoryArrayDma2{};
+            std::array<decltype(linkMemoryArrayDma1), 2> linkMemoryArray{ linkMemoryArrayDma1, linkMemoryArrayDma2 };
+#else
+            std::array<decltype(linkMemoryArrayDma1), 1> linkMemoryArray{ linkMemoryArrayDma1 };
```

Comment:
This allocates twice as much data as you think.

Once for linkMemoryArrayDma1/2 and once for linkMemoryArray. That also means they are unrelated, memory location wise.

Is this the intention?

Reply:
I think you meant something like:

```suggestion
            using LinkedListArray = std::array<LinkedList, 8>;
#if defined(DMA2_Channel1) || defined(GPDMA2)
            std::array<LinkedListArray, 2> linkMemoryArray{  };
#else
            std::array<LinkedListArray, 1> linkMemoryArray{  };
#endif
```

Reply:
it can be simplified further, but this requires GCC11 as a minimum:

```suggestion
            using LinkMemoryarray = std::array<LinkedList, 8>;
            std::array linkMemoryArray{ 
                LinkMemoryarray{},
#if defined(DMA2_Channel1) || defined(GPDMA2)
                LinkMemoryarray{}
#endif
             };
```

Reply:
![image](https://github.com/user-attachments/assets/62d06594-4449-45cb-bcf8-874e60315a9b)

Reply:
GCC 7 is already our minimum, so this is not a limiting factor

Reply:
Thank you for raising this point. Indeed it is possible to improve the memory usage.
I want to suggest a solution a bit different, since this last one has the same problem from my last last PR on DMA: the compiler is not creating an array of arrays. 

After some tests, I came up with this solution:

```
using LinkMemoryarray = std::array<LinkedList, 8>;

#if defined(DMA2_Channel1) || defined(GPDMA2)
     std::array<LinkMemoryarray, 2> linkMemoryArray{ LinkMemoryarray{}, LinkMemoryarray{} };
#else
     std::array<LinkMemoryarray, 1> linkMemoryArray{ LinkMemoryarray{} };
#endif
```

Reply:
From godbolt, it "should" generate it: see [here](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:13,endLineNumber:21,positionColumn:13,positionLineNumber:21,selectionStartColumn:13,selectionStartLineNumber:21,startColumn:13,startLineNumber:21),source:'%23include+%3Cstdint.h%3E%0A%0A%23include+%3Carray%3E%0A%0A%23define+GPDMA2%0A%0Astruct+LinkedList+%7B%0A++++uint32_t+A%3B%0A++++uint32_t+B%3B%0A++++uint32_t+C%3B%0A++++uint32_t+D%3B%0A%7D%3B%0A%0Ausing+LinkMemoryarray+%3D+std::array%3CLinkedList,+8%3E%3B%0Aconstexpr+std::array+linkMemoryArray%7BLinkMemoryarray%7B%7D,%0A%23if+defined(DMA2_Channel1)+%7C%7C+defined(GPDMA2)%0A+++++++++++++++++++++++++++++++++++++LinkMemoryarray%7B%7D%0A%23endif%0A%7D%3B%0A%0Aint+main()+%7B%0A++++auto%26+first+%3D+linkMemoryArray%5B0%5D%3B%0A++++auto%26+second+%3D+linkMemoryArray%5B1%5D%3B%0A%0A++++static_assert(sizeof(linkMemoryArray)+%3D%3D+2+*+(sizeof(LinkedList)+*+8))%3B%0A%0A++++return+0%3B%0A%7D'),l:'5',n:'1',o:'C%2B%2B+source+%231',t:'0')),k:37.703311777385856,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:armg1230,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-O0+-std%3Dc%2B%2B17',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+ARM+GCC+12.3.0+(Editor+%231)',t:'0')),k:28.963354889280815,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'ARM+gcc+10.3.1+(2021.10+none)',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+ARM+GCC+12.3.0+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)